scripts/verify.py

```diff
@@ -0,0 +1,184 @@
+        gdb.setR('R0', state.cpu.R0)
+
+    # Ignore Linux kernel helpers
+    if (state.cpu.PC >> 16) == 0xffff:
+        in_helper = True
+        return
+
+    # If we executed a few instructions of a helper, we need to sync Manticore's
+    # state to GDB as soon as we stop executing a helper.
+    if in_helper:
+        for reg in state.cpu.canonical_registers:
+            if reg.endswith('psr'):
+                continue
+            # Don't sync pc
+            if reg == 'pc':
+                continue
+            gdb.setR(reg, state.cpu.read_register(reg))
+        in_helper = False
+
+    if cmp_regs(state.cpu, should_print=True):
+        state.abandon()
+
+def sync_svc(state):
+    '''
+    Mirror some service calls in manticore. 
+    '''
+    name = linux_syscalls.armv7[syscall]
+    try:
+        # Make sure mmap returns the same address
+        if 'mmap' in name:
+            returned = gdb.getR('R0')
+            state.cpu.write_register('R0', returned)
+        if 'exit' in name:
+            return
+    except ValueError:
+        for reg in state.cpu.canonical_registers:
+            print '{}: {:x}'.format(reg, state.cpu.read_register(reg))
+        raise
+
+    logger.debug('Syscall: {} {}'.format(syscall, linux_syscalls.armv7[syscall]))
+    for i in range(4):
+        logger.debug("R{}: {:x}".format(i, gdb.getR('R%d'%i)))
+
+    state.cpu.memory.push_record_writes()
+
+def initialize(state):
+    '''
+    Synchronize the stack and register state
+    '''
+    gdb_regs = gdb.getCanonicalRegisters()
+    logger.debug("Copying {} bytes in the stack..".format(stack_top - state.cpu.SP))
+    state.cpu.SP = min(state.cpu.SP, gdb.getR('SP'))
+    #logger.debug("gdb val: %d", gdb.getR('SP')-state.cpu.SP)
+    for address in range(state.cpu.SP, stack_top):
+        b = gdb.getByte(address)
+        state.cpu.write_int(address, b, 8)
+    logger.debug("Done")
+    for reg in gdb_regs:
+        value = gdb_regs[reg]
+        if reg.endswith('psr'):
+            reg = 'apsr'
+        state.cpu.write_register(reg.upper(), value)
+
+
+def verify(argv):
+    logger.debug("Verifying program \"{}\"".format(argv))
+
+    # Address and stack_size are from linux.py
+    # TODO(yan): Refactor these constants into a reachable value in platform
+    qemu.start('arm', argv, va_size=stack_top, stack_size=stack_size)
+    gdb.start('arm', argv)
+
+    m = Manticore(argv[0], argv[1:])
+    m.verbosity = 3
+
+    @m.hook(None)
+    def on_instruction(state):
+        global initialized, last_instruction, syscall
+        # Initialize our state to QEMU's
+        if not initialized:
+            initialize(state)
+            initialized = True
+
+        if last_instruction is not None and last_instruction.address != state.cpu.PC:
+            on_after(state, last_instruction)
+
+        last_instruction = state.cpu.instruction
+
+        loc, instr = [x.strip() for x in gdb.getInstruction().split(':')]
+        mnemonic = instr.split('\t')[0]
```

Comment:
could we use `state.cpu.instruction.mnemonic` instead here? it should be the same as what gdb is about to exec?